File: Dsa-day65.py
from collections import deque, defaultdict
import bisect

class Router:
    def __init__(self, memoryLimit: int):
        self.limit = memoryLimit
        self.q = deque()
        self.seen = set()
        self.dest_map = defaultdict(list)

    def addPacket(self, source: int, destination: int, timestamp: int) -> bool:
        key = (source, destination, timestamp)
        if key in self.seen:
            return False
        if len(self.q) == self.limit:
            s, d, t = self.q.popleft()
            self.seen.remove((s, d, t))
            arr = self.dest_map[d]
            idx = bisect.bisect_left(arr, t)
            arr.pop(idx)
        self.q.append((source, destination, timestamp))
        self.seen.add(key)
        bisect.insort(self.dest_map[destination], timestamp)
        return True

    def forwardPacket(self) -> list:
        if not self.q:
            return []
        s, d, t = self.q.popleft()
        self.seen.remove((s, d, t))
        arr = self.dest_map[d]
        idx = bisect.bisect_left(arr, t)
        arr.pop(idx)
        return [s, d, t]

    def getCount(self, destination: int, startTime: int, endTime: int) -> int:
        arr = self.dest_map[destination]
        l = bisect.bisect_left(arr, startTime)
        r = bisect.bisect_right(arr, endTime)
        return r - l

File: Dsa-day65.java
import java.util.*;

class Router {
    int limit;
    Deque<int[]> q;
    Set<String> seen;
    Map<Integer, ArrayList<Integer>> destMap;

    public Router(int memoryLimit) {
        limit = memoryLimit;
        q = new ArrayDeque<>();
        seen = new HashSet<>();
        destMap = new HashMap<>();
    }

    public boolean addPacket(int source, int destination, int timestamp) {
        String key = source + "," + destination + "," + timestamp;
        if (seen.contains(key)) return false;
        if (q.size() == limit) {
            int[] pkt = q.poll();
            seen.remove(pkt[0] + "," + pkt[1] + "," + pkt[2]);
            ArrayList<Integer> arr = destMap.get(pkt[1]);
            int idx = Collections.binarySearch(arr, pkt[2]);
            arr.remove(idx < 0 ? -idx - 1 : idx);
        }
        q.offer(new int[]{source, destination, timestamp});
        seen.add(key);
        destMap.putIfAbsent(destination, new ArrayList<>());
        ArrayList<Integer> arr = destMap.get(destination);
        int idx = Collections.binarySearch(arr, timestamp);
        if (idx < 0) idx = -idx - 1;
        arr.add(idx, timestamp);
        return true;
    }

    public int[] forwardPacket() {
        if (q.isEmpty()) return new int[]{};
        int[] pkt = q.poll();
        seen.remove(pkt[0] + "," + pkt[1] + "," + pkt[2]);
        ArrayList<Integer> arr = destMap.get(pkt[1]);
        int idx = Collections.binarySearch(arr, pkt[2]);
        arr.remove(idx < 0 ? -idx - 1 : idx);
        return pkt;
    }

    public int getCount(int destination, int startTime, int endTime) {
        ArrayList<Integer> arr = destMap.getOrDefault(destination, new ArrayList<>());
        int l = Collections.binarySearch(arr, startTime);
        if (l < 0) l = -l - 1;
        int r = Collections.binarySearch(arr, endTime);
        if (r < 0) r = -r - 2;
        else {
            while (r + 1 < arr.size() && arr.get(r + 1) == endTime) r++;
        }
        return r - l + 1 < 0 ? 0 : r - l + 1;
    }
}
